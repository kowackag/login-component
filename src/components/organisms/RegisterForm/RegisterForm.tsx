import { InputField } from "@/components/molecules/InputField/InputField";
import { Button } from "@/components/atoms/Button/Button";

import { RegisterFormType } from "@/pages/RegisterPage/RegisterPage";

export const RegisterForm = ({ onSubmit, register, errors }: RegisterFormType) => {
	const {
		name: emailInputName,
		onChange: onEmailInputChange,
		onBlur: onEmailInputBlur,
		ref: emailInputRef,
	} = register("email");
	const {
		name: passwordInputName,
		onChange: onPasswordInputChange,
		onBlur: onPasswordInputBlur,
		ref: passwordInputRef,
	} = register("password");

	return (
		<form onSubmit={onSubmit} data-testid="register-form">
			<InputField
				name={emailInputName}
				label="Email"
				onChange={onEmailInputChange}
				onBlur={onEmailInputBlur}
				inputRef={emailInputRef}
				error={errors?.email?.message}
			/>
			<InputField
				name={passwordInputName}
				label="Password"
				type="password"
				onChange={onPasswordInputChange}
				onBlur={onPasswordInputBlur}
				inputRef={passwordInputRef}
				error={errors?.password?.message}
			/>
			<Button text="sign up" type="submit" />
		</form>
	);
};
