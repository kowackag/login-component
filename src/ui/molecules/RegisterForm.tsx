import { useForm } from "react-hook-form";
import { zodResolver } from "@hookform/resolvers/zod";

import { schema } from "@/views/authSchema";
import { ConfirmButton } from "@ui/atoms/ConfirmButton";

import { FormDataType } from "@/utils/shared.types";
import { FormField } from "@ui/molecules/FormField";

export const RegisterForm = () => {
	const {
		register,
		formState: { errors },
		handleSubmit,
		reset,
	} = useForm<Zod.output<typeof schema>>({
		resolver: zodResolver(schema),
		defaultValues: {
			email: "",
			password: "",
			saved: false,
		},
	});
	const onSubmit = (data: FormDataType) => {
		console.log(data);
		reset();
	};
	return (
		<form onSubmit={handleSubmit(onSubmit)}>
			<FormField name="email" label="Email" register={register} error={errors.email?.message} />
			<FormField
				name="password"
				label="Password"
				type="password"
				register={register}
				error={errors.password?.message}
			/>
			<ConfirmButton>Sign up</ConfirmButton>
		</form>
	);
};
