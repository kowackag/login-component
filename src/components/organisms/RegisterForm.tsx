import { useForm } from "react-hook-form";
import { zodResolver } from "@hookform/resolvers/zod";

import { schema } from "@/pages/authSchema";
import { ConfirmButton } from "@/components/atoms/ConfirmButton";

import { FormDataType } from "@/utils/shared.types";
import { InputField } from "@/components/molecules/InputField";

export const RegisterForm = () => {
	const {
		register,
		formState: { errors },
		handleSubmit,
	} = useForm<Zod.output<typeof schema>>({
		resolver: zodResolver(schema),
		defaultValues: {
			email: "",
			password: "",
			saved: false,
		},
	});
	const onSubmit = (data: FormDataType) => console.log(data);

	return (
		<form onSubmit={handleSubmit(onSubmit)}>
			<InputField name="email" label="Email" register={register} error={errors.email?.message} />
			<InputField
				name="password"
				label="Password"
				type="password"
				register={register}
				error={errors.password?.message}
			/>
			<ConfirmButton>Sign up</ConfirmButton>
		</form>
	);
};
