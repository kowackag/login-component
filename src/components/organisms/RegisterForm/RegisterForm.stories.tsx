import type { Meta } from "@storybook/react";
import { useForm } from "react-hook-form";
import { fn } from "@storybook/test";

import { RegisterForm } from "./RegisterForm";
import { RegisterFormDataTypes } from "@/pages/RegisterPage/RegisterPage";

const meta: Meta<typeof RegisterForm> = {
	component: RegisterForm,
	decorators: [
		(Story) => (
			<div style={{ width: "360px" }}>
				<Story />
			</div>
		),
	],
};

export default meta;

const onSubmit = fn();
const errors = {
	email: { message: "it is required", type: "required" },
	password: { message: "some error", type: "required" },
};

export const Default = () => {
	const { register } = useForm<RegisterFormDataTypes>();

	return <RegisterForm register={register} errors={errors} onSubmit={onSubmit} />;
};
