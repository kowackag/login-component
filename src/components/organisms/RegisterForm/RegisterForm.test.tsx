import { expect, it, describe, vi } from "vitest";
import { render, screen } from "@testing-library/react";
import { RegisterForm } from "./RegisterForm";
import { FieldErrors, UseFormRegister, FieldValues } from "react-hook-form";
type RegisterFieldsTypes = { email: string; password: string };

describe("RegisterForm", () => {
	it("RegisterForm exists", async () => {
		const mockFn = vi.fn();
		const register: UseFormRegister<RegisterFieldsTypes> = vi.fn((name) => {
			return {
				name: name as FieldValues["name"],
				onChange: mockFn,
				onBlur: mockFn,
				ref: mockFn,
			};
		});
		render(<RegisterForm onSubmit={mockFn} register={register} />);

		const form = screen.getByTestId("register-form");
		expect(form).toBeInTheDocument();
	});

	it("has appropriate numbers of children", async () => {
		const mockFn = vi.fn();
		const register: UseFormRegister<RegisterFieldsTypes> = vi.fn((name) => {
			return {
				name: name as FieldValues["name"],
				onChange: mockFn,
				onBlur: mockFn,
				ref: mockFn,
			};
		});
		render(<RegisterForm onSubmit={mockFn} register={register} />);
		const form = screen.getByTestId("register-form");
		expect(form.children.length).toBe(3);
	});

	it("shows appropriate error numbers and errors message", async () => {
		const mockFn = vi.fn();
		const register: UseFormRegister<RegisterFieldsTypes> = vi.fn((name) => {
			return {
				name: name as FieldValues["name"],
				onChange: mockFn,
				onBlur: mockFn,
				ref: mockFn,
			};
		});
		const errors: FieldErrors<RegisterFieldsTypes> = {
			email: {
				message: "Invalid email",
				type: "invalid form",
			},
                  password: {
                        message: "Invalid password",
				type: "invalid form",
                  }
		};
		render(<RegisterForm onSubmit={mockFn} register={register} errors={errors} />);
		const errorList = screen.getAllByTestId("error");
		expect(errorList).toHaveLength(2);
		expect(errorList[0]).toHaveTextContent("Invalid email");
            expect(errorList[1]).toHaveTextContent("Invalid password");
	});
	it("has button type submit", async () => {
		const mockFn = vi.fn();
            const register: UseFormRegister<RegisterFieldsTypes> = vi.fn((name) => {
			return {
				name: name as FieldValues["name"],
				onChange: mockFn,
				onBlur: mockFn,
				ref: mockFn,
			};
		});
		render(<RegisterForm onSubmit={mockFn} register={register} />);

		const button = screen.getByRole("button", { name: /sign up/i });
		expect(button).toHaveAttribute("type", "submit");
	});

	it("renders correctly", () => {
		const mockFn = vi.fn();
		const register: UseFormRegister<RegisterFieldsTypes> = vi.fn((name) => {
			return {
				name: name as FieldValues["name"],
				onChange: mockFn,
				onBlur: mockFn,
				ref: mockFn,
			};
		});
		const { container } = render(<RegisterForm onSubmit={mockFn} register={register} />);
		expect(container).toMatchSnapshot();
	});
});
