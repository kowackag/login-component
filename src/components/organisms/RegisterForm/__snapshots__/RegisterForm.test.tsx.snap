// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`RegisterForm > renders correctly 1`] = `
<div>
  <form
    data-testid="register-form"
  >
    <div
      class="relative py-3"
    >
      <label
        class="inline-block font-primaryMedium"
        for="email"
      >
        Email
      </label>
      <input
        class="w-full rounded-md border-2 border-neutral-300 bg-transparent px-2 leading-9 outline-none"
        id="email"
        name="email"
        type="text"
      />
    </div>
    <div
      class="relative py-3"
    >
      <label
        class="inline-block font-primaryMedium"
        for="password"
      >
        Password
      </label>
      <input
        class="w-full rounded-md border-2 border-neutral-300 bg-transparent px-2 leading-9 outline-none"
        id="password"
        name="password"
        type="password"
      />
    </div>
    <button
      class="my-2 w-full rounded-lg bg-bright-rose text-center font-primaryMedium text-sm uppercase leading-10 tracking-widest text-neutral-100 shadow-btn"
      type="submit"
    >
      sign up
    </button>
  </form>
</div>
`;
